nexus-sdk/src/main/java/io/nexusrpc/OperationHandler.java

```diff
@@ -0,0 +1,228 @@
+    private final @Nullable String asyncOperationId;
+
+    private StartResult(@Nullable R syncResult, @Nullable String asyncOperationId) {
+      this.syncResult = syncResult;
+      this.asyncOperationId = asyncOperationId;
+    }
+
+    /** Whether this start result is synchronous or asynchronous. */
+    public boolean isSync() {
+      return asyncOperationId == null;
+    }
+
+    /**
+     * The synchronous result. This can be null if the result is actually null or it is an
+     * asynchronous operation.
+     */
+    public @Nullable R getSyncResult() {
+      return syncResult;
+    }
+
+    /** The asynchronous operation ID. This will be null if the operation result is synchronous. */
+    public @Nullable String getAsyncOperationId() {
+      return asyncOperationId;
+    }
+  }
+
+  /** Context for {@link #start}. */
+  interface StartContext {
+    /** Service name. */
+    String getService();
+
+    /** Operation name. */
+    String getOperation();
+
+    /** Headers. Key accessing on this map is case-insensitive. */
+    Map<String, String> getHeaders();
+
+    /**
+     * Optional callback for asynchronous operations to deliver results to. If this is present and
+     * the implementation is an asynchronous operation, the implementation should ensure this
+     * callback is invoked with the result upon completion.
+     */
+    @Nullable String getCallbackUrl();
+
+    /** Headers to use on the callback if {@link #getCallbackUrl} is used. */
+    Map<String, String> getCallbackHeaders();
+
+    /** Unique request identifier from the caller to be used for deduplication. */
+    String getRequestId();
+  }
+
+  /** Context for {@link #fetchResult}. */
+  interface FetchResultContext {
+    /** Service name. */
+    String getService();
+
+    /** Operation name. */
+    String getOperation();
+
+    /** Operation ID. */
+    String getOperationId();
+
+    /**
+     * Optional timeout for how long the user want to wait on the result.
+     *
+     * <p>If this value is null, the result or {@link OperationStillRunningException} should be
+     * returned/thrown right away. If this value is present, the fetch result call should try to
+     * wait up until this duration or until an implementer chosen maximum, whichever ends sooner,
+     * before returning the result or throwing {@link OperationStillRunningException}.
+     */
+    @Nullable Duration getTimeout();
+
+    /** Headers. Key accessing on this map is case-insensitive. */
+    Map<String, String> getHeaders();
+  }
+
+  /** Context for {@link #fetchInfo}. */
+  interface FetchInfoContext {
+    /** Service name. */
+    String getService();
+
+    /** Operation name. */
+    String getOperation();
+
+    /** Operation ID. */
+    String getOperationId();
+
+    /** Headers. Key accessing on this map is case-insensitive. */
+    Map<String, String> getHeaders();
+  }
+
+  /** Context for {@link #cancel}. */
+  interface CancelContext {
```

Comment:
Is there a reason these are interfaces and not just classes?  Looking at the Nexus Go SDK all the hanler options are just structs?

Reply:
Yes, IMO the impl of these are going to need to be defined outside of this class (but can be in this same project, unsure how this will shake out once handler impl comes about). I also wonder if the properties that have getters here would need to be accessible mutably by the inner workings outside of this package. I think these interfaces give the safest separation between contract and impl in this case.

Reply:
Yeah agree with avoiding thread locals here.

Do you plan to use the `context for timeouts and cancellation of all requests? I see the `FetchResultContext ` has a timeout but no other context does? In principle all of these requests are RPCs and should have a timeout and be cancellable right? `gRPC Java` uses a context for these things and the Java SDK does respect the gRPC context when making an RPC.

Personally I would separate the context and the options into two separate parameters if you can have a common `context` type for all operations.

Reply:
I _think_ only `fetchResult` is the only operation that should not be immediate and therefore should be cancellable. But I admit I have not put much thought into cancellation. I am curious what cancellation approach we even want to offer here, it'd need to be designed. But I agree, if cancellation is a common thing for all and in other ways if there is a common context for all, then splitting to have a generic operation context and specific method infos works for me too.

I am not sure we have to design worker shutdown cancellation at this time. There is of course user/server-sent cancellation which calls `cancel` and should interrupt any `fetchResult` calls. But I think we may be able to defer how worker cancellation is implemented here at least at first.

Reply:
These methods willRPC's though right? So a user could want to specify a timeout for the RPC. In the nexus go SDK the obvious approach is to use the  [context](https://github.com/nexus-rpc/sdk-go/blob/6447c77cfa89cebd3e425f1c8d364897748a6d3e/nexus/handle.go#L171), but with Java we will have to come up with our own pattern I think.

>I am not sure we have to design worker shutdown interruption at this time.

Yeah not saying we need to design that here. My main concern is taking interfaces here instead of POJO, what flexibility is this buying us and why do we need it in Java Nexus SDK but not the Go Nexus SDK?

Reply:
> with Java we will have to come up with our own pattern I think.

Agreed. I will set aside some time to research how, say, gRPC Java interrupts/notifies gRPC implementation calls. Also, can you confirm that if Temporal Java activities don't heartbeat, there is no other interruption mechanism (e.g. thread interrupt)?

> what flexibility is this buying us and why do we need it in Java Nexus SDK but not the Go Nexus SDK?

Only because there was no Java context already I figured there was no value in any shared context object (as opposed to Go where there is a shared context object). If we can design interruption simply right now and do want a shared context, I will move these to POJOs (though it will clutter up this file a bit, but meh, still worth having as inner classes).

Reply:
>Also, can you confirm that if Temporal Java activities don't heartbeat, there is no other interruption mechanism (e.g. thread interrupt)?

Correct, I would like to add an opt in option to interrupt the activity, but currently it is up to users to check the activity info for the deadline and cancel themselves.

> If we can design interruption simply right now and do want a shared context,

I think how we do interruption is a part of the API design we need to think about because it dictates these operation method signatures.

Reply:
> I think how we do interruption is a part of the API design we need to think about because it dictates these operation method signatures.

Concur, ideas? gRPC Java has `io.grpc.Context` which has methods for knowing it was cancelled, why it was cancelled, and registering cancellation callbacks. Think we should do something similar? Then we can have one shared context interface and I will make the method-specific information basic POJOs.

Reply:
gRPC Java passes `io.grpc.Context` through a thread local, but we don't need to copy that hehe. Might be worthing looking at the HTTP libraries in Java and seeing how they represent timeouts and cancellation.

Reply:
Yeah, I'm taking some of that context idea but making it more explicit.